### ISSUE TITLE
clamp makes grad None

### ISSUE THREAD

**Issue by theFool32** [pytorch/pytorch#6991]
## Issue description

It seems `torch.clamp` makes `grad` of a tensor with `requires_grad=True` `None`
I don't know whether it is a bug or I misuse it.
Thanks.

## Code example
```
# snippet1
a = torch.rand(1, requires_grad=True)
a = torch.clamp(a, 0, 1)
z = 2 * a
z.backward()
print(a.requires_grad) # True
print(a.grad) # None

# snippet2
a = torch.rand(1, requires_grad=True)
a = torch.clamp(a, 0, 1)
z = 2 * a
z.backward()
print(a.requires_grad) # True
print(a.grad) # tensor([ 2.])

```

## System Info
Collecting environment information...
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Arch Linux
GCC version: (GCC) 7.3.1 20180406
CMake version: version 3.10.3

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy (1.14.2)
[pip3] torch (0.4.0)
[pip3] torchvision (0.2.0)
[conda] Could not collect

**Comment by fmassa** [comment]
The reason why in your first snippet the grad is `None` is because you are modifying the leaf variable `a` and you overwrite it with the result of `a.clamp`. This means that `a` now is an intermediate variable, and the gradient gets freed once it's not needed anymore.
You should instead do:
```python
a = torch.rand(1, requires_grad=True)
b = torch.clamp(a, 0, 1)
z = 2 * b
z.backward()
print(a.requires_grad) # True
print(a.grad) # None
```
